timemachine/cpp/src/bd_exchange_move.hpp

```diff
@@ -47,20 +45,30 @@ template <typename RealType> class BDExchangeMove : public Mover {
     DeviceBuffer<RealType> d_log_sum_exp_after_;  // [2]
     DeviceBuffer<int>
         d_samples_; // where the indices to sample a molecule come from, currently fixed to a single sample
-    DeviceBuffer<RealType> d_quaternions_;  // Normal noise for uniform random rotations
-    DeviceBuffer<RealType> d_translations_; // Uniform noise for translation + the check
+    DeviceBuffer<RealType> d_quaternions_; // Normal noise for uniform random rotations
     DeviceBuffer<size_t> d_num_accepted_;
     DeviceBuffer<int> d_target_mol_atoms_;
     DeviceBuffer<int> d_target_mol_offsets_;
     DeviceBuffer<__int128> d_intermediate_sample_weights_;
+    DeviceBuffer<RealType> d_sample_noise_;          // Noise to use for selecting molecules
+    DeviceBuffer<RealType> d_sampling_intermediate_; // [num_target_mols_] Intermediate buffer for weighted sampling
 
     curandGenerator_t cr_rng_;
 
     void compute_initial_weights(const int N, double *d_coords, double *d_box, cudaStream_t stream);
 
     void compute_incremental_weights(
-        const int N, const bool scale, double *d_coords, double *d_box, RealType *d_quaternions, cudaStream_t stream);
-
+        const int N,
+        const bool scale,
+        const double *d_box,
+        const double *d_coords,
+        const RealType *d_quaternions,
+        const RealType *d_translations,
+        cudaStream_t stream);
+
+private:
+    // Keep this private, so we can realloc the buffer in the targeted case
+    DeviceBuffer<RealType> d_translations_; // Uniform noise for translation + the check
```

Comment:
I think in C++, the way private inheritance works is that the base class's `d_translations_` is still *present* (and taking up gpu memory), just in-accessible. Not sure if that's desirable if we never use it.

Reply:
(Just kind of ugly from an inheritance PoV)

Reply:
Yes, this does take up memory (https://stackoverflow.com/questions/20859384/does-an-instantiation-of-a-derived-class-allocate-memory-for-private-members-of)

The other solution would be to call `realloc`, removing the warning we have right now for realloc when the array is not empty. I had previously tried to override it and kept getting segfaults. Didn't get to the bottom of it for time reasons, but could re-investigate. This is probably going to become an issue when we batch so probably good to fix now

Reply:
Another (maybe bad) idea: could you use a protected constructor in `BDExchangeMove` that doesn't initialize `d_translations_`, and use this in `TIBDExchangeMove`?

Reply:
Will play around with both approaches and see which seems cleaner.

Reply:
Ended up using the constructor as the virtual method doesn't work in constructors: https://isocpp.org/wiki/faq/strange-inheritance#calling-virtuals-from-ctors

https://github.com/proteneer/timemachine/pull/1234/commits/f839e788c2dbf7706efd1c7adc47fd97699db200

Reply:
ah yeah, I forgot about that gotcha re: virtuals in constructors, I hate c++